### PR TITLE
[stdlib] Rewrite UTF8 decoding

### PR DESCRIPTION
**Pull request by PatrickPijnappel**
#### What's in this pull request?

Expanding upon #1477, this is a thorough rewrite of UTF8.decode() for optimization, code simplification and clarity.
- Merged maximal subpart determination and conversion to code point with `_isValidUTF8()` (renamed to `_decodeOne()`), significantly reducing total amount of code and eliminating unnecessary branching.
- Very significant ASCII decoding speed-up by skipping the buffer.
- Overall performance improvement, branch elimination and code simplification.
- Added several clarification comments throughout. 
- Expanded validation testing to count sequence lengths and maximal subparts.
- Validated against original implementation for all inputs up to 4 code units (i.e. all 2^32 possibilities).

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by dabrahams** [comment]
@PatrickPijnappel Hi; can you please merge master into your branch and resolve conflicts?  Thanks.

File: stdlib/public/core/Unicode.swift [hunk 1]
```diff
@@ -59,7 +59,9 @@ public protocol UnicodeCodecType {
   /// Because of buffering, it is impossible to find the corresponding position
   /// in the generator for a given returned `UnicodeScalar` or an error.
   ///
-  /// - parameter next: A generator of code units to be decoded.
+  /// - Parameter next: A generator of code units to be decoded.  Repeated
+  /// calls to this method on the same instance should always reuse the same
+  /// generator.  Failing to do so will result in undefined behavior.
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: Please indent following lines of a bullet by two spaces.
What do you mean by “reuse” here?  Would “pass” be a better word?

I don’t think you mean undefined behavior; that actually includes memory and type-safety violations.  I think you mean the results are unspecified.  I think you also want to say that the generator or copies thereof shouldn’t be used for anything else between calls to this method.  This seems like a complicated-enough restriction to warrant an example of correct usage.

File: stdlib/public/core/Unicode.swift [hunk 2]
```diff
@@ -71,51 +73,131 @@ public protocol UnicodeCodecType {
 
 /// A codec for [UTF-8](http://www.unicode.org/glossary/#UTF_8).
 public struct UTF8 : UnicodeCodecType {
+  // See Unicode 8.0.0, Ch 3.9, UTF-8.
+  // http://www.unicode.org/versions/Unicode8.0.0/ch03.pdf
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: Is this addition actually useful?

Comment (PatrickPijnappel) [inline stdlib/public/core/Unicode.swift]: Personally it would have helped me when reading the original code, but maybe that's just me ;).

**Comment by PatrickPijnappel** [comment]
@dabrahams Resolved conflicts and fixed documentation.

These methods would be much more self-explanatory if we just had:
``` swift
static func decode<S where S.Generator.Element == CodeUnit>(sequence: S) -> SomeSequence<UnicodeScalar>
static func encode<S where S.Generator.Element == UnicodeScalar>(sequence: S) -> SomeSequence<CodeUnit>
```
Is there a specific reason for the current design to be preferable?

**Comment by dabrahams** [comment]
@swift-ci Please test

**Comment by PatrickPijnappel** [comment]
@dabrahams Please test!

**Comment by gottesmm** [comment]
@swift-ci Please test

File: stdlib/public/core/StringUTF8.swift [hunk 3]
```diff
@@ -159,7 +159,9 @@ extension String {
       /// True iff the index is at the end of its view or if the next
       /// byte begins a new UnicodeScalar.
       internal var _isOnUnicodeScalarBoundary : Bool {
-        return UTF8._isValidUTF8(UInt32(truncatingBitPattern: _buffer)) || _isAtEnd
+        let buffer = UInt32(truncatingBitPattern: _buffer)
+        let (codePoint, _) = UTF8._decodeOne(buffer)
+        return codePoint != nil || _isAtEnd
       }
```
Comment (rintaro) [inline stdlib/public/core/StringUTF8.swift]: @PatrickPijnappel 
Just out of curiosity, why `return !UTF8.isContinuation(UTF8.CodeUnit(truncatingBitPattern: buffer))` is insufficient?

Comment (PatrickPijnappel) [inline stdlib/public/core/StringUTF8.swift]: That would change the behavior slightly w.r.t. the original version, as not all non-continuation code units are valid starting code units (e.g. `0xff`). Though the difference is somewhat inconsequential as `_buffer` should be well-formed, and in general it's not clear from the documentation what the handling of mal-formed sequences should be.

Comment (rintaro) [inline stdlib/public/core/StringUTF8.swift]: Thanks for clarifying :)

**Comment by PatrickPijnappel** [comment]
@gottesmm I can't reproduce these test failures on my machine (using `build-script -RcT`) and they don't seem to be related to the PR. Any suggestions?

**Comment by dabrahams** [comment]
@swift-ci Please test

**Comment by dabrahams** [comment]
on Mon Mar 14 2016, Patrick Pijnappel <notifications-AT-github.com> wrote:

> @gottesmm I can't reproduce these test failures on my machine (using
> `build-script -RcT`) and they don't seem to be related to the PR. Any
> suggestions?

I'm re-testing.  Upstream seems to be healthy now, but maybe this test
ran at a time that it was broken.

-- 
-Dave

**Comment by PatrickPijnappel** [comment]
@dabrahams Ok, now OS X fails but Linux passes. Still seems unrelated to the changes though.

**Comment by dabrahams** [comment]
This looks like it was a failure in upstream: https://ci.swift.org/job/oss-swift_tools-RA_stdlib-DA_test-simulator/790/consoleFull#-1806112821fca400bf-2f4a-462e-b517-e058d770b2d7

File: stdlib/public/core/Unicode.swift [hunk 4]
```diff
@@ -63,13 +64,17 @@ public protocol UnicodeCodec {
   /// Because of buffering, it is impossible to find the corresponding position
   /// in the iterator for a given returned `UnicodeScalar` or an error.
   ///
-  /// - parameter next: An iterator over the code units to be decoded.
+  /// - Parameter next: A generator of code units to be decoded.  Repeated
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: Please change "generator" to "iterator"

File: stdlib/public/core/Unicode.swift [hunk 5]
```diff
@@ -78,51 +83,133 @@ public protocol UnicodeCodec {
...
   public init() {}
 
-  /// Lookahead buffer used for UTF-8 decoding.  New bytes are inserted at LSB,
-  /// and bytes are read at MSB.
-  internal var _decodeLookahead: UInt32 = 0
+  /// Lookahead buffer used for UTF-8 decoding.  New bytes are inserted at MSB,
+  /// and bytes are read at LSB.  Note that we need to use a buffer, because
+  /// in case of invalid subsequences we sometimes don't know whether we should
+  /// consume a certain byte before looking at it.
+  var _decodeBuffer: UInt32 = 0
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: please use explicit "internal" for non-public decls.

File: stdlib/public/core/Unicode.swift [hunk 6]
```diff
@@ -140,260 +227,61 @@ public struct UTF8 : UnicodeCodec {
     let bit1 = (lut1 >> index) & 1
 
     switch (bit1, bit0) {
-    case (0, 0): // 2-byte sequence.
+    case (0, 0): // 2-byte sequence, [ XXX XXX CU1 CU0 ].
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: `[ XXX XXX CU1 CU0 ]` doesn't make sense to me.  Everywhere else we've been using X's for bits.  What do they mean here?

File: stdlib/public/core/Unicode.swift [hunk 7]
```diff
@@ -464,7 +352,11 @@ public struct UTF16 : UnicodeCodec {
   /// Because of buffering, it is impossible to find the corresponding position
   /// in the iterator for a given returned `UnicodeScalar` or an error.
   ///
-  /// - parameter next: An *iterator* over the code units to be decoded.
+  /// - Parameter next: A generator of code units to be decoded.  Repeated
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: s/generator/iterator/

File: stdlib/public/core/Unicode.swift [hunk 8]
```diff
@@ -847,19 +742,19 @@ extension UTF16 {
...
+        _memcpy(
+          dest: UnsafeMutablePointer(destination),
+          src: UnsafeMutablePointer(source),
+          size: UInt(count) * UInt(strideof(U.self)))
+      }
+      else {
+        for i in 0..<count {
+          let u16 = T._toUTF16CodeUnit((source + i).pointee)
+          (destination + i).pointee = U._fromUTF16CodeUnit(u16)
+        }
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: Please fix the indentation here (-2 spaces).  In general it's a good idea to check your diffs for whitespace changes, which only introduce churn.

File: stdlib/public/core/Unicode.swift [hunk 9]
```diff
@@ -829,7 +724,7 @@ extension UTF16 {
     _precondition(width(x) == 2)
     return UTF16.CodeUnit(
       (x.value - 0x1_0000) & (((1 as UInt32) << 10) - 1)
-    ) + 0xDC00
+      ) + 0xDC00
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: Please unindent 2 spaces

File: stdlib/public/core/Unicode.swift [hunk 10]
```diff
@@ -873,35 +768,49 @@ extension UTF16 {
...
+            count += width(us)
+          case .emptyInput:
+            break loop
+          case .error:
+            if !repairingIllFormedSequences {
+              return nil
+            }
+            isAscii = false
+            count += width(UnicodeScalar(0xfffd))
+          }
```
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: Please unindent 2 spaces

**Comment by dabrahams** [comment]
We're really close to being able to accept this one!

File: stdlib/public/core/Unicode.swift (hunk 4 above)
Comment (PatrickPijnappel) [inline stdlib/public/core/Unicode.swift]: Actually I also found a few stray cases of generator in the rest of the stdlib, see #1705.

File: stdlib/public/core/Unicode.swift (hunk 6 above)
Comment (PatrickPijnappel) [inline stdlib/public/core/Unicode.swift]: `XXX` denotes one byte here, to show the layout of the bytes in the buffer. I've used `x` (lowercase) for bits. Maybe a single `X` per byte, or any other suggestions?

File: stdlib/public/core/Unicode.swift (hunk 8 above)
Comment (PatrickPijnappel) [inline stdlib/public/core/Unicode.swift]: Thanks, I'll be sure to check next time!

**Comment by PatrickPijnappel** [comment]
@dabrahams Cool :).

File: stdlib/public/core/Unicode.swift (hunk 6 above)
Comment (dabrahams) [inline stdlib/public/core/Unicode.swift]: I don't know; I can't tell what you're trying to express here.  presumably CU1 and CU0 are additional bytes, so it looks like a 4-byte sequence.

Comment (PatrickPijnappel) [inline stdlib/public/core/Unicode.swift]: It's the byte layout of buffer, i.e. the code units (CU0 & CU1) are in the lower 2 bytes and the upper two (the XXXs) are garbage.

**Comment by gribozavr** [comment]
@swift-ci Please test

**Pull request merged**